cmd/tink-server/main.go

```diff
@@ -130,48 +135,59 @@ func NewRootCommand(config *DaemonConfig, logger log.Logger) *cobra.Command {
 				config.PGPassword,
 				config.PGSSLMode,
 			)
-
-			dbCon, err := sql.Open("postgres", connInfo)
+			database, err := internal.SetupPostgres(connInfo, config.OnlyMigration, logger)
 			if err != nil {
 				return err
 			}
-			tinkDB := db.Connect(dbCon, logger)
-
 			if config.OnlyMigration {
-				logger.Info("Applying migrations. This process will end when migrations will take place.")
-				numAppliedMigrations, err := tinkDB.Migrate()
+				return nil
+			}
+
+			var (
+				grpcOpts    []grpc.ServerOption
+				certPEM     []byte
+				certModTime *time.Time
+			)
+			if config.TLS {
+				certsDir := os.Getenv("TINKERBELL_CERTS_DIR")
+				if certsDir == "" {
+					certsDir = filepath.Join("/certs", config.Facility)
+				}
+				var cert *tls.Certificate
+				cert, certPEM, certModTime, err = grpcserver.GetCerts(certsDir)
 				if err != nil {
 					return err
 				}
-				logger.With("num_applied_migrations", numAppliedMigrations).Info("Migrations applied successfully")
-				return nil
+				grpcOpts = append(grpcOpts, grpc.Creds(credentials.NewServerTLSFromCert(cert)))
 			}
 
-			numAvailableMigrations, err := tinkDB.CheckRequiredMigrations()
+			tinkAPI, err := server.NewDBServer(
+				logger,
+				database,
+				server.WithCerts(*certModTime, certPEM),
+			)
 			if err != nil {
 				return err
 			}
-			if numAvailableMigrations != 0 {
-				logger.Info("Your database schema is not up to date. Please apply migrations running tink-server with env var ONLY_MIGRATION set.")
-			}
 
-			grpcConfig := &grpcServer.ConfigGRPCServer{
-				Facility:      config.Facility,
-				TLSCert:       "insecure",
-				GRPCAuthority: config.GRPCAuthority,
-				DB:            tinkDB,
-			}
-			if config.TLS {
-				grpcConfig.TLSCert = config.TLSCert
+			// Start the gRPC server in the background
+			addr, err := grpcserver.SetupGRPC(
+				ctx,
+				tinkAPI,
+				config.GRPCAuthority,
+				grpcOpts,
+				errCh)
+			if err != nil {
+				return err
 			}
-			cert, modT := grpcServer.SetupGRPC(ctx, logger, grpcConfig, errCh)
+			logger.With("address", addr).Info("started listener")
 
-			httpConfig := &httpServer.Config{
+			httpConfig := &httpserver.Config{
 				HTTPAuthority: config.HTTPAuthority,
-				CertPEM:       cert,
-				ModTime:       modT,
+				CertPEM:       certPEM,
+				ModTime:       *certModTime,
 			}
-			httpServer.SetupHTTP(ctx, logger, httpConfig, errCh)
+			httpserver.SetupHTTP(ctx, logger, httpConfig, errCh)
 
```

Comment:
This could really benefit from using https://github.com/oklog/run. Unnecessary for this PR, if you get time it would help clean up the subsequent error channel code that does static draining of the channel.